rivOrtho.R

```diff
@@ -0,0 +1,362 @@
+  b[j] = -999
+  
+  
+  
+  
+  
+  
+  #### handle river segment ends: 
+  
+  # recalculate jumps, this time over a single nextdoor neighbor vertices:
+  p1x = x[-l]
+  p1y = y[-l]
+  p2x = x[-1]
+  p2y = y[-1]
+  
+  p1 = cbind(p1x, p1y)
+  p2 = cbind(p2x, p2y)
+  
+  sd = distGeo(p1, p2)
+  sj = which(sd > res*2)+1
+  
+  
+  # insert NAs at start, end, and jump in vector:
+  for (i in rev(1:length(sj))){
+    x = insertRow(x, NA, sj[i])
+    y = insertRow(y, NA, sj[i])
+    b = insertRow(b, NA, sj[i])
+    w = insertRow(w, NA, sj[i])
+  }
+  
+  x = c(NA, x, NA)
+  y = c(NA, y, NA)
+  b = c(NA, b, NA)
+  w = c(NA, w, NA)
+  
+  
+  # get bounds of -999 values:
+  jNA = which(is.na(b))
+  
+  
+  closeL = jNA[-1] - 1
+  closeR = jNA[-length(jNA)] + 1
+  farL = jNA[-1] - floor(n/2)
+  farR = jNA[-length(jNA)] + floor(n/2)
+  
+  
+  
+  # use a linearly shrinking window to calculate bearing at ends of vectors:
+  for (i in 1:(length(jNA)-1)){
+    
+    fL = farL[i]:closeL[i]
+    rL = closeR[i]:farR[i]
+    
+    
+    for (ii in 1:length(fL)){
+      
+      # calculate all points on left sides of jumps:
+      L = c((fL[ii]-floor(n/2)), closeL[i])
+      b[fL[ii]] = bearing(cbind(x[L[1]], y[L[1]]), cbind(x[L[2]], y[L[2]]))
+
+      # handle all points on right sides of vectors:
+      R = c(closeR[i], (rL[ii]+floor(n/2)))
+      b[rL[ii]] = bearing(cbind(x[R[1]], y[R[1]]), cbind(x[R[2]], y[R[2]]))
+
+    }
+  }
+  
+  # occationally, there are a situations where the GRWL
+  # centerline is clipped such that there is only 1 segment, 
+  # thus introducing NANs into the bearing calculation above.
+  # fill these NANs in with a bear of 90. 
+  b[which(is.nan(b))] = 90
+  b[jNA] = NA
+  
+
+  
+  #which(b==-999) # there should be no -999s 
+  
+  
+  
+  
+  #########################################################################
+  # interpolate/extrapolate any missing width values:
+  
+  wNA = which(w<30)
+  if (length(wNA) > 0){
+    jw = c(0, which(wNA[-1]-wNA[-length(wNA)]>1), length(wNA))
+    
+    #jwNA = is.na(w[wNA[jw]+1])
+    
+    win = 5 # number of measurements to use to fill in missing widths
+    
+    # for each block of missing width values:
+    for (i in 1:(length(jw)-1)){
+      
+      # find out what is on either side of the block of missing widths:
+      wB = wNA[(jw[i]+1):jw[i+1]]
+      lwB = length(wB)
+      
+      L = w[wB[1]-1]
+      R = w[wB[lwB]+1]
+      
+      # if a block of missing widths does not contain a jump on either end,
+      
+      if (!is.na(L) & !is.na(R)){
+        # linear interpolation:
+        m = (w[wB[lwB]+1]-w[wB[1]-1])/(lwB+1)
+        w[wB] = m * (1:lwB) +  w[wB[1]-1]
+        
+        # interpolate between missing width values with a cubic spline:
+        #spx = c(c(1:win), c(1:win)+lwB)
+        #spy = c(w[c((wB[1]-win):(wB[1]-1))], w[c((wB[lwB]+1):(wB[lwB]+win))])
+        #spf = splinefun(spx, spy)
+        #w[wB] = spf(c((win+1):(win+lwB)))

```

Comment:
Either add to comment to explain why you are leaving this here, for example "The following shows how to interpolate between missing width values using a cubic spline if the data set requires it". Or, if its not needed anymore then remove it.